contracts/smart-account/modules/BatchedSessionRouterModule.sol

```diff
@@ -0,0 +1,190 @@
+            (bytes, address)
+        );
+
+        /*
+        struct SessionCallData {
+            uint48 validUntil;
+            uint48 validAfter;
+            address sessionValidationModule;
+            bytes sessionKeyData;
+            bytes32[] merkleProof;
+            bytes callSpecificData;
+        }
+        */
+
+        // parse the signature to get the array of required parameters
+        (
+            address sessionKeyManager,
+            uint48[] memory validUntil,
+            uint48[] memory validAfter,
+            address[] memory sessionValidationModule,
+            bytes[] memory sessionKeyData,
+            bytes32[][] memory merkleProof,
+            bytes[] memory callSpecificData,
+            bytes memory sessionKeySignature
+        ) = abi.decode(
+                moduleSignature,
+                (
+                    address,
+                    uint48[],
+                    uint48[],
+                    address[],
+                    bytes[],
+                    bytes32[][],
+                    bytes[],
+                    bytes
+                )
+            );
+
+        // can I use struct instead so don't have to check lengths
+
+        // check lengths of arrays
+        require(
+            validUntil.length == validAfter.length,
+            "SR Invalid data provided"
+        );
+        require(
+            validUntil.length == sessionValidationModule.length,
+            "SR Invalid data provided"
+        );
+        require(
+            validUntil.length == sessionKeyData.length,
+            "SR Invalid data provided"
+        );
+        require(
+            validUntil.length == merkleProof.length,
+            "SR Invalid data provided"
+        );
+
+        address recovered = ECDSA.recover(
+            ECDSA.toEthSignedMessageHash(
+                keccak256(abi.encodePacked(userOpHash, sessionKeyManager))
+            ),
+            sessionKeySignature
+        );
+
+        uint48 earliestValidUntil = type(uint48).max;
+        uint48 latestValidAfter;
+
+        // iterate over batched operations
+        for (uint i; i < sessionValidationModule.length; ) {
+            // validate the sessionKey
+            // sessionKeyManager reverts if something wrong
+            ISessionKeyManager(sessionKeyManager).validateSessionKey(
+                userOp.sender,
+                validUntil[i],
+                validAfter[i],
+                sessionValidationModule[i],
+                sessionKeyData[i],
+                merkleProof[i]
+            );
+
+            (address sessionKey, , , ) = abi.decode(
+                sessionKeyData[i],
+                (address, address, address, uint256)
+            );
+
+            // compare if userOp was signed with the proper session key
+            if (recovered != sessionKey) return SIG_VALIDATION_FAILED;
+
+            // parse userOp.calldata to get calldatas for every specific operation
+            (
```

Comment:
Given that this is independent of `i`, shouldn't this be moved to outside the for loop to save gas?

Reply:
nope, because `sessionKey` depends on `i`

Reply:
Ah inwas referring to the next section, where we decode userop.calldata to get the the batch

Reply:
ah, I see. Yes, this definitely should be moved out of the loop. Thank you!